<?xml version="1.0"?>

<robot name="miarobot" xmlns:xacro="http://ros.org/wiki/xacro">
  <xacro:include filename="$(find miarobot_description)/urdf/model.gazebo"/>
  <xacro:include filename="$(find miarobot_description)/urdf/model_materials.xacro"/>
  <xacro:include filename="$(find miarobot_description)/urdf/model.wheels.xacro"/>
 

  <xacro:property name="use_lidar" value="$(arg use_lidar)"/>
  <xacro:property name="kinectMass" value="0.25"></xacro:property>
  <xacro:property name="kinectLength" value="0.25"></xacro:property>
  <xacro:property name="kinectWidth" value="0.04"></xacro:property>
  <xacro:property name="kinectHeight" value="0.03"></xacro:property>

  <!-- ======================================= BASE LINK ======================================= -->
  
  <link name="base_link">
    <visual>
      <geometry>
          <mesh filename="package://miarobot_description/meshes/miarobot_colored.dae"/>
      </geometry>
      <material name="blue"/>
      <origin rpy="0 0 0" xyz="0 0 0"/> 
    </visual>

    <collision>
      <geometry>
        <mesh filename="package://miarobot_description/meshes/miarobot.dae"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/> 
    </collision>

    <inertial>
      <mass value="5"/>
      <inertia ixx="10" ixy="0.0" ixz="0.0" iyy="10" iyz="0.0" izz="10"/>
    </inertial>

  </link>

<!-- ======================================== WHEELS AND JOINTS ========================================= -->

  <!-- Instantiating wheels -->
  <xacro:miarobot_wheel wheel_suffix="left_front" joint_origin_xyz="0.15 0.15 0.04" >
  </xacro:miarobot_wheel>
  <xacro:miarobot_wheel wheel_suffix="left_back" joint_origin_xyz="-0.15 0.15 0.04">
  </xacro:miarobot_wheel>
  <xacro:miarobot_wheel wheel_suffix="right_front" joint_origin_xyz="0.15 -0.15 0.04">
  </xacro:miarobot_wheel>
  <xacro:miarobot_wheel wheel_suffix="right_back" joint_origin_xyz="-0.15 -0.15 0.04">
  </xacro:miarobot_wheel>

  <!-- Kinect Camera -->
  <link name="kinect_camera_link">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="${kinectWidth} ${kinectLength} ${kinectHeight}"/>
      </geometry>
    </collision>

    <!-- <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="${kinectWidth} ${kinectLength}  ${kinectHeight}"/>
      </geometry>
      <material name="green"/>
    </visual> -->
    <inertial>
      <mass value="${kinectMass}" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <box_inertia m="${kinectMass}" x="{kinectLength}" y="${kinectWidth}" z="${kinectHeight}" />
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

  <joint name="kinect_camera_joint" type="fixed">
    <axis xyz="0 0 0" />
    <origin xyz="0.16 0 0.185" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="kinect_camera_link"/>
  </joint>

  <!-- LIDAR -->
  <xacro:if value="${use_lidar}">
    <link name="laser">
      <collision>
        <origin xyz="0 0 0.26" rpy="0 0 0"/>
        <geometry>
          <box size="0.1 0.1 0.1"/>
        </geometry>
      </collision>

      <inertial>
        <mass value="1e-5" />
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
      </inertial>
    </link>

    <joint name="hokuyo_joint" type="fixed">
      <axis xyz="0 1 0" />
      <origin xyz="0 0 0.26" rpy="0 0 0"/>
      <parent link="base_link"/>
      <child link="laser"/>
    </joint>
  </xacro:if>

  <!-- IMU Sensor -->

  <link name="imu_link">
    <!-- <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.001 0.001 0.001"/>
      </geometry>
    </collision> -->

    <visual>
      <geometry>
        <box size="0.001 0.001 0.001"/>
      </geometry>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>
  
  <joint name="imu_joint" type="fixed">
    <axis xyz="0 1 0" />
    <origin xyz="0 0 0.18" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="imu_link"/>
  </joint>

  <link name="base_ir_back">
    <collision>
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>
    <visual>
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </visual>
    <inertial>
      <mass value="1e-5" />
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0"      izz="1e-6" />
    </inertial>
  </link>

  <joint name="ir_back_joint" type="fixed">
    <axis xyz="0 1 0" />
    <origin xyz="0 0 0.26" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="base_ir_back"/>
  </joint>



  <!-- Ultrasonic Front -->
<link name="ultrasonic_front_link">
  <visual>
    <geometry><box size="0.01 0.01 0.01"/></geometry>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </visual>
  <inertial><mass value="1e-5"/></inertial>
</link>
<joint name="ultrasonic_front_joint" type="fixed">
  <parent link="base_link"/>
  <child link="ultrasonic_front_link"/>
  <origin xyz="0.2 0 0.1" rpy="0 0 0"/> <!-- Adjust position -->
</joint>

<!-- Ultrasonic Back -->
<link name="ultrasonic_back_link">
  <visual>
    <geometry><box size="0.01 0.01 0.01"/></geometry>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </visual>
  <inertial><mass value="1e-5"/></inertial>
</link>
<joint name="ultrasonic_back_joint" type="fixed">
  <parent link="base_link"/>
  <child link="ultrasonic_back_link"/>
  <origin xyz="-0.2 0 0.1" rpy="0 0 3.1416"/> <!-- Facing backwards -->
</joint>

<!-- Ultrasonic Left -->
<link name="ultrasonic_left_link">
  <visual>
    <geometry><box size="0.01 0.01 0.01"/></geometry>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </visual>
  <inertial><mass value="1e-5"/></inertial>
</link>
<joint name="ultrasonic_left_joint" type="fixed">
  <parent link="base_link"/>
  <child link="ultrasonic_left_link"/>
  <origin xyz="0 0.15 0.1" rpy="0 0 1.5708"/> <!-- Facing left -->
</joint>


  <!-- ==================================================== Arm ===================================================== -->
  <!--Base-->
  <link name="arm_base">
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="1e-4" ixy="0.0" ixz="0.0" iyy="1e-4" iyz="0.0" izz="1e-4"/>
    </inertial>

    <visual>
      <origin xyz="0 0.0 0.002" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.05"/>
      </geometry>
      <material name="blue"/>
    </visual>

    <collision>
      <geometry>
        <cylinder radius="0.02" length="0.1"/>
      </geometry>
    </collision>
  </link>

  <joint name="arm_base_joint" type="fixed">
    <origin xyz="0.0 0.09 0.25" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="arm_base"/>
  </joint>

  <!--Height joint-->
  <link name="joint1_link">
    <inertial>
      <mass value="0.01"/>
      <inertia ixx="1e-5" ixy="0.0" ixz="0.0" iyy="1e-5" iyz="0.0" izz="1e-5"/>
    </inertial>
    <visual name="">
      <origin xyz="0.0 0.0 0.08" rpy="0.0 0 0.0"/>
      <geometry>
        <cylinder radius="0.03" length="0.2"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <collision>
      <origin xyz="0.0 0.0 0.09" rpy="0.0 1.5705 0.0"/>
      <geometry>
        <cylinder radius="0.03" length="0.2"/>
      </geometry>
    </collision>
  </link>

  <joint name="joint1" type="prismatic">
    <origin xyz="0.0 0.0 0.048" rpy="0.0 0.0 0.0"/>
    <parent link="arm_base"/>
    <child link="joint1_link"/>
    <axis xyz="0.0 0.0 1"/>
    <limit lower="0.0" upper="0.4" effort="10" velocity="1"/>
  </joint>
  

  <!--rotation in xy Joint-->
  <link name="joint2_link">
    <inertial>
      <origin xyz="0.15 0.0 0.0" rpy="0.0 0.0 0.0"/>
      <mass value="0.005"/>
      <inertia ixx="1e-6" ixy="0.0" ixz="0.0" iyy="1e-6" iyz="0.0" izz="1e-6"/>
    </inertial>
    <visual>
      <origin xyz="0.085 0.0 -0.01" rpy="0.0 0.0 0.0"/> <!--extend forward-->
      <geometry>
        <box size="0.2 0.03 0.03"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0.085 0.0 0.0" rpy="0.0 0.0 0.0"/>
      <geometry>
        <box size="0.2 0.03 0.03"/>
      </geometry>
    </collision>
  </link>

  <joint name="link2_joint" type="revolute">
    <origin xyz="0.0 0.0 0.195" rpy="0.0 0.0 0.0"/>
    <parent link="joint1_link"/>
    <child link="joint2_link"/>
    <axis xyz="0.0 0.0 1"/>     <!--rotation around z-->
    <limit lower="-1.8" upper="1.8" effort="5" velocity="1.0"/>
  </joint>

  <!--================================================ Gripper =============================================== -->
  <link name="gripper_base">
    <inertial>
      <mass value="0.01"/>
      <inertia ixx="1e-6" ixy="0" ixz="0.0" iyy="1e-6" iyz="0.0" izz="1e-6"/>
    </inertial>
    <visual>
      <origin xyz="-0.01 0.0 -0.01" rpy="0.0 0.0 0.0"/> 
      <geometry>
        <cylinder radius="0.015" length="0.03"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <collision>
      <origin xyz="-0.01 0.0 -0.006" rpy="0.0 0.0 0.0"/>
      <geometry>
        <cylinder radius="0.015" length="0.03"/>
      </geometry>
    </collision>
  </link>

  <joint name="gripper_base_joint" type="fixed">
    <origin xyz="0.198 0.0 0" rpy="0.0 0.0 0.0"/>
    <parent link="joint2_link"/>
    <child link="gripper_base"/>
  </joint>

  <!--left_part-->
  <link name="left_finger">
    <inertial>
      <mass value="0.001"/>
      <inertia ixx="1e-6" ixy="0" ixz="0.0" iyy="1e-6" iyz="0.0" izz="1e-6"/>
    </inertial>
    <visual>
      <geometry>
        <box size="0.05 0.01 0.02"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <geometry>
        <box size="0.05 0.01 0.02"/>
      </geometry>
    </collision>
  </link>

  <joint name="left_finger_joint" type="prismatic">
    <origin xyz="0.015 -0.02 -0.012" rpy="0.0 0.0 0.0"/>
    <parent link="gripper_base"/>
    <child link="left_finger"/>
    <axis xyz="0.0 1 0.0"/>
    <limit lower="0.0" upper="0.012" effort="5" velocity="0.1"/>
  </joint>

  <!--right_part-->
  <link name="right_finger">
    <inertial>
      <mass value="0.001"/>
      <inertia ixx="1e-6" ixy="0" ixz="0.0" iyy="1e-6" iyz="0.0" izz="1e-6"/>
    </inertial>
    <visual>
      <geometry>
        <box size="0.05 0.01 0.02"/>
      </geometry>
      <material name="red"/>
    </visual>
    <collision>
      <geometry>
        <box size="0.05 0.01 0.02"/>
      </geometry>
    </collision>
  </link>

  <joint name="right_finger_joint" type="prismatic">
    <origin xyz="0.015 0.02 -0.012" rpy="0.0 0.0 0.0"/>
    <parent link="gripper_base"/>
    <child link="right_finger"/>
    <axis xyz="0.0 -1 0.0"/>
    <limit lower="0.0" upper="0.012" effort="5" velocity="0.1"/>
  </joint>



 </robot>
